#                                     Survival Prediction on Titanic Dataset 

My [Git Hub ID](https://github.com/sreerag050897)<br/>
My [LinkdeIn ID](https://www.linkedin.com/in/sreeragpe/)

I have downloaded the titanic data set from [here](https://www.kaggle.com/competitions/titanic/data).<br/>
My objective is to build a Classifiction ML model.<br/>
Here I have used Logistic Regression and K-Nearest Neighbours for the same

## Importing the necessary data processing packages

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df=pd.read_csv(r'train.csv')

### Now I am performing some EDA on the data set

In [6]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [7]:
df.shape

(891, 12)

In [10]:
df.isnull().values.any()

True

In [11]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [13]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

### It is pretty evident that you are machine can't learn from categorical data.<br/> So I have used the "get_dummies" attributes of Pandas libraries to change the categorical value of the features "Embarked" and "Sex"

In [15]:
df=pd.get_dummies(data=df,prefix='Embarked',columns=['Embarked'])

In [21]:
df=pd.get_dummies(data=df,columns=['Sex'],drop_first=True)

In [25]:
df['Sex']=df['Sex_male']

In [27]:
df.drop('Sex_male',axis=1,inplace=True)

In [28]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name   Age  SibSp  Parch  \
0                              Braund, Mr. Owen Harris  22.0      1      0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0      1      0   
2                               Heikkinen, Miss. Laina  26.0      0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0      1      0   
4                             Allen, Mr. William Henry  35.0      0      0   
..                                                 ...   ...    ...    ...   
886                              Montvila, Rev. Juozas  27.0      0      0   
887                       Graham, Miss. Margaret Edith  19.0      0      0   
888           Johnston, Miss. Catherine Helen "Carrie"   NaN      1      2   
889                              Behr, Mr. Karl Howell  26.0      0      0   
890                                Dooley, Mr. Patrick  32.0      0      0   

               Ticket     Fare Cabin  Embarked_C  Embarked_Q  Embarked_S  Sex  
0           A/5 21171   7.2500   NaN           0           0           1    1  
1            PC 17599  71.2833   C85           1           0           0    0  
2    STON/O2. 3101282   7.9250   NaN           0           0           1    0  
3              113803  53.1000  C123           0           0           1    0  
4              373450   8.0500   NaN           0           0           1    1  
..                ...      ...   ...         ...         ...         ...  ...  
886            211536  13.0000   NaN           0           0           1    1  
887            112053  30.0000   B42           0           0           1    0  
888        W./C. 6607  23.4500   NaN           0           0           1    0  
889            111369  30.0000  C148           1           0           0    1  
890            370376   7.7500   NaN           0           1           0    1  

[891 rows x 14 columns]

#### Here I split the data to 2 data set X and Y<br/>
X intenting the data set without Survived column where Y represents the survived column

In [29]:
x=df.drop('Survived',axis=1)

In [40]:
y=df.loc[:,'Survived']

In [41]:
x.drop('S')

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: Survived, Length: 891, dtype: int64

#### Now I am removing some features from the dependent data set that seems no effect on the prediction.<br/> It is always good to remove unwanted features as it will affect the model's accuracy to great extend at times.

#### I am droping the features such as Cabin, Ticket, Name, PassengerID

In [42]:
x.drop(['Cabin','Ticket','Name','PassengerId'],axis=1,inplace=True)

In [43]:
x

Pclass   Age  SibSp  Parch     Fare  Embarked_C  Embarked_Q  Embarked_S  \
0         3  22.0      1      0   7.2500           0           0           1   
1         1  38.0      1      0  71.2833           1           0           0   
2         3  26.0      0      0   7.9250           0           0           1   
3         1  35.0      1      0  53.1000           0           0           1   
4         3  35.0      0      0   8.0500           0           0           1   
..      ...   ...    ...    ...      ...         ...         ...         ...   
886       2  27.0      0      0  13.0000           0           0           1   
887       1  19.0      0      0  30.0000           0           0           1   
888       3   NaN      1      2  23.4500           0           0           1   
889       1  26.0      0      0  30.0000           1           0           0   
890       3  32.0      0      0   7.7500           0           1           0   

     Sex  
0      1  
1      0  
2      0  
3      0  
4      1  
..   ...  
886    1  
887    0  
888    0  
889    1  
890    1  

[891 rows x 9 columns]

Above you can see I have dropped the mentioned features

###### It is to be noted that there must not be any null values whilst training a model. 

In [45]:
x.isnull().values.any()

True

In [46]:
x.isnull().sum()

Pclass          0
Age           177
SibSp           0
Parch           0
Fare            0
Embarked_C      0
Embarked_Q      0
Embarked_S      0
Sex             0
dtype: int64

#### There are 177 null values for Age column.<br/> It can either be dropped or filled using any feature Engineering techniques.

#### I have imputed the null values using the mean value

In [50]:
x=x.fillna(x['Age'].mean())

In [53]:
x.isnull().values.any()

False

##### Now I am splitting the data to train and test sets using SKlearn library

In [54]:
from sklearn.model_selection import train_test_split

In [55]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.25,random_state=0)

##### I am trying to build a model using Logistic regression

In [56]:
from sklearn.linear_model import LogisticRegression

In [57]:
model=LogisticRegression()

In [58]:
model.fit(x_train,y_train)

C:\Users\sreer\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [60]:
model.score(x_train,y_train)

0.8008982035928144

In [61]:
model.score(x_test,y_test)

0.7982062780269058

## As you can see I got the model with 80 percentgae accuracy with Logistic Regression

### Now I am trying K-Nearest Neighbors algorithm to build a model

In [62]:
from sklearn.neighbors import KNeighborsClassifier

In [76]:
model1=KNeighborsClassifier(n_neighbors=15)

In [77]:
model1.fit(x_train,y_train)

KNeighborsClassifier(n_neighbors=15)

In [78]:
model1.score(x_train,y_train)

0.7290419161676647

In [79]:
model1.score(x_test,y_test)

0.7533632286995515

## Here I got the model with 75 percentgae accuracy